refactor "Run Main" to open a new terminal to compile the flix program

manoj2601 opened the pull request:
fix #102 

For now, two methods are added:
Method 1: It simply checks whether there is an active terminal with the name `flix` in vscode window. If no, then it creates a new terminal named `flix` and compiles the flix program in that terminal. else it compiles in the existing terminal with the name `flix`.

Method 2: If there are no active terminals, it simply creates a new terminal named `flix` and compiles the program in it. If there are active terminals, then it opens a drop-down window in vscode to select a terminal to compile.
a sample drop-down to select a terminal:
![Screenshot from 2021-06-18 10-01-17](https://user-images.githubusercontent.com/45541010/122507586-15e72c00-d01e-11eb-8e0a-d59be9e83bde.png)

magnus-madsen commented:
Can you paste a screenshot of what it looks like when the command is executed?

magnus-madsen commented on client/src/handlers/handlers.ts:
```diff
@@ -14,6 +14,86 @@ export function makeHandleRunJob (
+export async function compileInTerminal() {
+    
+    const files = await vscode.workspace.findFiles(FLIX_GLOB_PATTERN)
+    let cmd = "java -jar "+flixFileLocation;
```
Need to ensure that all paths are escaped correctly. Maybe there is a package or function in VSCode?

magnus-madsen commented:
I have added 4 commands:

![image](https://user-images.githubusercontent.com/2287953/122545251-ecce9780-d02d-11eb-9ef4-9829f8562e4a.png)

manoj2601 commented:
updated according to these 4 commands.

magnus-madsen commented:
Thanks!

You should be able to test with the newest nightly:

https://flix.dev/nightly/

Ill take a pass over the code soon.

magnus-madsen commented on client/src/extension.ts:
```diff
@@ -105,8 +105,12 @@ export async function activate (context: vscode.ExtensionContext, launchOptions:
+  registerCommand('flix.cmdRunMain', handlers.RunInExistingTerminalWithoutArg)
+  registerCommand('flix.runMainWithArgs', handlers.RunInExistingTerminalWithArg)
+  registerCommand('flix.runMainNewTerminal', handlers.RunInNewTerminalWithoutArg)
+  registerCommand('flix.runMainNewTerminalWithArgs', handlers.RunInNewTerminalWithArg)
```
Let's keep the names consistent, i.e.:

RunInExistingTerminalWithoutArg -> runMain
RunInExistingTerminalWithArg -> runMainWithArgs
RunInNewTerminalWithoutArg -> runMainNewTerminal
RunInNewTerminalWithArg -> runMainNewTerminalWithArgs

magnus-madsen replied:
Just noticed that you need `cmd` in front too!

magnus-madsen commented on client/src/extension.ts:
```diff
@@ -105,8 +105,12 @@ export async function activate (context: vscode.ExtensionContext, launchOptions:
+  registerCommand('flix.runMainNewTerminal', handlers.RunInNewTerminalWithoutArg)
+  registerCommand('flix.runMainNewTerminalWithArgs', handlers.RunInNewTerminalWithArg)
+  
+//   registerCommand('flix.cmdRunMain', handlers.makeHandleRunJobWithProgress(client, outputChannel, jobs.Request.cmdRunMain, 'Running Main'))
```
This can be deleted.

magnus-madsen commented on client/src/extension.ts:
```diff
@@ -200,6 +204,7 @@ async function startSession (context: vscode.ExtensionContext, launchOptions: La
 
   // Wait until we're sure flix exists
   const flixFilename = await ensureFlixExists({ globalStoragePath, workspaceFolders, shouldUpdateFlix: launchOptions.shouldUpdateFlix })
+  handlers.setFlixFileName(flixFilename) 
```
I think this is dangerous, because now this depends on things being loaded in the right order.

I think this should rather be set on the context object or something like that.

magnus-madsen commented on client/src/handlers/handlers.ts:
```diff
@@ -14,6 +14,189 @@ export function makeHandleRunJob (
+*/
+export async function RunInNewTerminalWithoutArg() {
+    let terminal = ensureNewFlixTerminal()
+    let cmd = "java -jar "+flixFileLocation
```
The `flixFileLocation` also needs to be quoted/escape (because the jar could be stored in a folder with a space in it).

magnus-madsen commented on client/src/handlers/handlers.ts:
```diff
@@ -14,6 +14,189 @@ export function makeHandleRunJob (
+*/
+export async function RunInExistingTerminalWithoutArg() {
+    let terminal = ensureFlixTerminal()
+    let cmd = "java -jar "+flixFileLocation
```
The `flixFileLocation` also needs to be quoted/escape (because the jar could be stored in a folder with a space in it).

magnus-madsen commented on client/src/handlers/handlers.ts:
```diff
@@ -14,6 +14,189 @@ export function makeHandleRunJob (
+ * @return void
+*/
+async function passArgs(terminal:vscode.Terminal) {
+    let cmd = "java -jar "+flixFileLocation
```
The `flixFileLocation` also needs to be quoted/escape (because the jar could be stored in a folder with a space in it).

magnus-madsen commented on client/src/handlers/handlers.ts:
```diff
@@ -14,6 +14,189 @@ export function makeHandleRunJob (
+async function takeInputFromUser() {
+    const input = await vscode.window.showInputBox({
+        prompt: "Enter the space separated arguments",
+        placeHolder: "Foo bar",
```
I think it should just be empty or omitted. We don't want users to have to erase "Foo bar" every time.

manoj2601 replied:
Now it is replaced with `arg0 arg1 arg2 ...`.  Also, users don't need to erase it. This is just a shadow in the input box. Like this:
![Screenshot from 2021-06-21 20-04-16](https://user-images.githubusercontent.com/45541010/122779831-e4938800-d2cb-11eb-9c69-e0f6a58ce836.png)

magnus-madsen commented on client/src/handlers/handlers.ts:
```diff
@@ -14,6 +14,189 @@ export function makeHandleRunJob (
+*/
+async function takeInputFromUser() {
+    const input = await vscode.window.showInputBox({
+        prompt: "Enter the space separated arguments",
```
Enter the program arguments (separated by spaces).

magnus-madsen commented on client/src/handlers/handlers.ts:
```diff
@@ -14,6 +14,189 @@ export function makeHandleRunJob (
+async function getFiles() {
+    const files = await vscode.workspace.findFiles(FLIX_GLOB_PATTERN)
+    let cmd = ""
+    for (let index = 0; index < files.length ; index++)
```
You should check out array join :)

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/join

manoj2601 replied:
resolved

magnus-madsen commented on client/src/handlers/handlers.ts:
```diff
@@ -14,6 +14,189 @@ export function makeHandleRunJob (
+    const files = await vscode.workspace.findFiles(FLIX_GLOB_PATTERN)
+    let cmd = ""
+    for (let index = 0; index < files.length ; index++)
+        cmd += " \""+files[index].path+"\"";
```
So this escaping is not really sufficient. At least in theory the filename could contain a " character. 

Maybe we should just use:

https://www.npmjs.com/package/shell-quote

and call that with an array of the right arguments.  I think that would work very well.

magnus-madsen commented on client/src/handlers/handlers.ts:
```diff
@@ -14,6 +14,189 @@ export function makeHandleRunJob (
+        passCommandToTerminal(cmd, terminal)  
+    }
+    else {
+        vscode.window.showErrorMessage("Process cancelled")
```
I think it is okay to cancel and then we simply don't have to show anything.

magnus-madsen commented on client/src/handlers/handlers.ts:
```diff
@@ -14,6 +14,189 @@ export function makeHandleRunJob (
+function ensureNewFlixTerminal() {
+    const count = countFlixTerminals()
+    if(count == 0)
+        return vscode.window.createTerminal(`flix`);
```
I think the first terminal should be called `flix-0`, not just `flix`.

magnus-madsen commented on client/src/handlers/handlers.ts:
```diff
@@ -14,6 +14,189 @@ export function makeHandleRunJob (
+ * @return vscode.Terminal
+*/
+function ensureNewFlixTerminal() {
+    const count = countFlixTerminals()
```
Would it make sense to instead maintain an integer ourselves?

manoj2601 replied:
Assume the user creates 5 terminals (flix-0, flix-1, ...., flix-4) and then deletes all. Now if we maintain an integer, the next created terminal will be flix-5 although there are no active flix terminals.

mlutze replied:
What if the user creates 2 terminals, then deletes the first one, then creates a new one? Then there would be two terminals called `flix-1`, right?

manoj2601 replied:
> What if the user creates 2 terminals, then deletes the first one, then creates a new one? Then there would be two terminals called `flix-1`, right?

yes

magnus-madsen replied:
So, If I start closing some terminals it could be pretty hard to know which terminal the program is currently running in?

manoj2601 replied:
> So, If I start closing some terminals it could be pretty hard to know which terminal the program is currently running in?

:thumbsup: updated. maintained an integer that keeps track of the total number of flix terminals.

magnus-madsen commented on client/src/handlers/handlers.ts:
```diff
@@ -14,6 +14,189 @@ export function makeHandleRunJob (
+*/
+function ensureFlixTerminal() {
+    const activeTerminals = vscode.window.terminals
+    for (let i = 0; i < activeTerminals.length; i++) {
```
In modern JS you can write:

```
let colors = ['red', 'green', 'blue'];
for (const color of colors){
    console.log(color);
}
```

see https://stackoverflow.com/questions/3010840/loop-through-an-array-in-javascript

magnus-madsen replied:
same comment applies to countFlixTerminals

magnus-madsen commented on client/src/handlers/handlers.ts:
```diff
@@ -14,6 +14,189 @@ export function makeHandleRunJob (
 }
 
+let flixFileLocation
+const FLIX_GLOB_PATTERN = '**/*.flix'
```
This is already defined somewhere else. Maybe we can reuse that?

magnus-madsen commented:
Forgot to say, if you merge/rebase onto master, you can avoid those debug errors.

magnus-madsen commented:
Also make sure to test it against the latest nightly.

manoj2601 commented:
![Screenshot from 2021-06-21 18-13-46](https://user-images.githubusercontent.com/45541010/122763926-8612dd80-d2bc-11eb-99e9-ce41365de77b.png)

Can someone suggest some good choices for prompt and placeholder while taking input from the user as shown in the screenshot?

magnus-madsen commented:
> ![Screenshot from 2021-06-21 18-13-46](https://user-images.githubusercontent.com/45541010/122763926-8612dd80-d2bc-11eb-99e9-ce41365de77b.png)
> 
> Can someone suggest some good choices for prompt and placeholder while taking input (arguments for the flix program) from the user as shown in the screenshot?

"Enter arguments separated by spaces".

And I think arg0 arg1 and so forth- what you have- is fine.

magnus-madsen commented on client/src/handlers/handlers.ts:
```diff
@@ -14,6 +18,211 @@ export function makeHandleRunJob (
+async function getFiles() {
+    const files = await vscode.workspace.findFiles(FLIX_GLOB_PATTERN)
+    let paths = files.map(x => x.path)
+    let cmd = require('shell-quote').quote(paths)
```
You need to run `npm install shell-quote` and let it update `package.json`. Otherwise it will not build on other machines.

magnus-madsen replied:
Also, instead of a require there should be an import at the top of the file.

manoj2601 replied:
updated

magnus-madsen commented on client/src/extension.ts:
```diff
@@ -14,13 +14,13 @@ import initialiseState from './services/state'
 import { callResolversAndEmptyList } from './services/timers'
 
-const _ = require('lodash/fp')
+export const _ = require('lodash/fp')
```
We should not re-export lodash. Intead you can import it in the other file.

magnus-madsen replied:
(The other exports are fine)

magnus-madsen commented on client/src/handlers/handlers.ts:
```diff
@@ -14,6 +19,211 @@ export function makeHandleRunJob (
+ *
+ * @return vscode.Terminal
+*/
+function ensureFlixTerminal() {
```
Rename to `getFlixTerminal`

magnus-madsen commented on client/src/handlers/handlers.ts:
```diff
@@ -14,6 +19,211 @@ export function makeHandleRunJob (
+ *
+ * @return vscode.Terminal
+*/
+function ensureNewFlixTerminal() {
```
rename to `newFlixTerminal`.

magnus-madsen commented on client/src/handlers/handlers.ts:
```diff
@@ -14,6 +19,211 @@ export function makeHandleRunJob (
+*/
+async function passArgs(terminal:vscode.Terminal, flixFilename: string) {
+
+    let cmd = "java -jar "+ flixFilename + " "
```
Refactoring:

Instead of calling quote on `flixFilename` and building a string here, just build one array:

`['java', '-jar',flixFileName]`

and then push the arguments on to the array (if any). 

I.e. the point is to **completely avoid string concatenation**.

magnus-madsen commented on client/src/handlers/handlers.ts:
```diff
@@ -14,6 +19,211 @@ export function makeHandleRunJob (
+    let input = await takeInputFromUser()
+    if(!(input == undefined || input.trim().length == 0)) {
+        cmd += " --args ";
+        cmd += quote([input]);
```
OK, except for the input, here you probably have to use string concatenation.

magnus-madsen commented on client/src/handlers/handlers.ts:
```diff
@@ -14,6 +19,211 @@ export function makeHandleRunJob (
+            const globalStoragePath = context.globalStoragePath
+            const workspaceFolders = _.map(_.flow(_.get('uri'), _.get('fsPath')), vscode.workspace.workspaceFolders)
+            const flixFileLocation = await ensureFlixExists({ globalStoragePath, workspaceFolders, shouldUpdateFlix: launchOptions.shouldUpdateFlix })
+            var flixFilename = quote([flixFileLocation])
```
See my comment above. 

Specifically I don't like that we call quote here, because it is easy to forget somewhere. Its better to always build a big array and then pass the whole thing to quote!

magnus-madsen commented on client/src/handlers/handlers.ts:
```diff
@@ -14,6 +19,211 @@ export function makeHandleRunJob (
+        return async function handler () {
+            const globalStoragePath = context.globalStoragePath
+            const workspaceFolders = _.map(_.flow(_.get('uri'), _.get('fsPath')), vscode.workspace.workspaceFolders)
+            const flixFileLocation = await ensureFlixExists({ globalStoragePath, workspaceFolders, shouldUpdateFlix: launchOptions.shouldUpdateFlix })
```
The above three lines seems to be copied in several functions. Create a helper function?

magnus-madsen reviewed (commented):
I added some comments. Overall I think the code quality has much improved 👍 .

My one major concern is still the excessive use of string concatenation. It is better to just build arrays with all the arguments and then use quote on the entire array.

magnus-madsen commented:
Other than the Windows error, the code looks ready 👍

magnus-madsen commented:
@manoj2601 I think instead of `.path` you just need to use `.fsPath`.

manoj2601 commented:
> @manoj2601 I think instead of `.path` you just need to use `.fsPath`.

updated.
It's strange that both functions `path` and `fspath` give the same output in ubuntu system but different in windows.
I have updated it and tested in both ubuntu and windows machine.

magnus-madsen commented:
Merged. Nice!
